File: study/dp/fib.py
memo = {0 : 0, 1 : 1}

def fib(n, memo):
    memo = memo or {}
    if n in memo:
        return memo[n]
    if n == 0:
        return 0
    elif n == 1:
        return 1    
    memo[n] = fib(n - 1, memo) + fib(n - 2, memo)
    return memo[n] 

for i in range(100):
    res = fib(i, memo)
    print(f"Result: {res}")

File: hw/act1.2/fib/fib.cpp
#include <iostream>
#include "HashTable.h"

int fib(int num, HashTable & memo) { 
    if (memo.search(num) != -1) { 
        return memo.search(num); 
    }
    if (num == 0) { 
        return 0; 
    }
    if (num == 1) { 
        return 1; 
    }
    memo.insert(num, fib(num - 1, memo) + fib(num - 2, memo));
    return memo.search(num);
}

int main() { 
    HashTable memo(10); 
    memo.insert(0, 0);
    memo.insert(1,1);
    
    for (int i = 0; i < 100; i++) { 
        std::cout << fib(i, memo) << std::endl; 
    }
    memo.displayTable();
}
